stl/inc/algorithm

```diff
@@ -2634,98 +2637,72 @@ namespace ranges {
     template <class _It, class _Ty>
     using fold_left_first_with_iter_result = in_value_result<_It, _Ty>;
 
-    template <class _It, class _Se, class _Ty, class _Fn>
-    _NODISCARD constexpr auto _Fold_left_with_iter_unchecked(_It _First, _Se _Last, _Ty _Init, _Fn _Func) {
-        _STL_INTERNAL_STATIC_ASSERT(input_iterator<_It>);
-        _STL_INTERNAL_STATIC_ASSERT(sentinel_for<_Se, _It>);
-        _STL_INTERNAL_STATIC_ASSERT(_Indirectly_binary_left_foldable<_Fn, _Ty, _It>);
-
-        using _Uty         = decay_t<invoke_result_t<_Fn&, _Ty, iter_reference_t<_It>>>;
-        using _Return_type = fold_left_with_iter_result<_It, _Uty>;
-        if (_First == _Last) {
-            return _Return_type{_STD move(_First), _Uty(_STD move(_Init))};
-        } else {
-            _Uty _Accum = _STD invoke(_Func, _STD move(_Init), *_First);
-            for (++_First; _First != _Last; ++_First) {
-                _Accum = _STD invoke(_Func, _STD move(_Accum), *_First);
-            }
-            return _Return_type{_STD move(_First), _STD move(_Accum)};
-        }
-    }
-
-    class _Fold_left_fn : private _Not_quite_object {
+    class _Fold_left_with_iter_fn : private _Not_quite_object {
     public:
         using _Not_quite_object::_Not_quite_object;
 
         template <input_iterator _It, sentinel_for<_It> _Se, class _Ty, _Indirectly_binary_left_foldable<_Ty, _It> _Fn>
         _NODISCARD constexpr auto operator()(_It _First, _Se _Last, _Ty _Init, _Fn _Func) const {
             _Adl_verify_range(_First, _Last);
-            return _Fold_left_with_iter_unchecked(_Unwrap_iter<_Se>(_STD move(_First)),
-                _Unwrap_sent<_It>(_STD move(_Last)), _STD move(_Init), _Pass_fn(_Func))
-                .value;
+            auto _UResult = _Fold_left_with_iter_unchecked(_Unwrap_iter<_Se>(_STD move(_First)),
+                _Unwrap_sent<_It>(_STD move(_Last)), _STD move(_Init), _Pass_fn(_Func));
+            _Seek_wrapped(_First, _STD move(_UResult.in));
+
+            using _Uty         = decay_t<invoke_result_t<_Fn&, _Ty, iter_reference_t<_It>>>;
+            using _Return_type = fold_left_with_iter_result<_It, _Uty>;
+            return _Return_type{_STD move(_First), _STD move(_UResult.value)};
         }
 
         template <input_range _Rng, class _Ty, _Indirectly_binary_left_foldable<_Ty, iterator_t<_Rng>> _Fn>
         _NODISCARD constexpr auto operator()(_Rng&& _Range, _Ty _Init, _Fn _Func) const {
-            return _Fold_left_with_iter_unchecked(
-                _Unwrap_range_iter<_Rng>(_RANGES begin(_Range)), _Uend(_Range), _STD move(_Init), _Pass_fn(_Func))
-                .value;
+            using _Uty         = decay_t<invoke_result_t<_Fn&, _Ty, range_reference_t<_Rng>>>;
+            using _Return_type = fold_left_with_iter_result<borrowed_iterator_t<_Rng>, _Uty>;
+            return static_cast<_Return_type>(
+                (*this)(_RANGES begin(_Range), _RANGES end(_Range), _STD move(_Init), _Pass_fn(_Func)));

```

Comment:
This now does a triple move; I meant more like "pass the wrapped iterator type to `_Fold_left_with_iter_unchecked`, so that it could do the `_Seek_wrapped`"; do you mind if I push the code I'd like?

Reply:
Sure

Reply:
how do you feel about the changes

Reply:
Great!
and I'm sorry, I totally misunderstood your previous comment 🤦‍♂️